core/src/main/java/org/jruby/specialized/RubyObjectSpecializer.java

```diff
@@ -0,0 +1,277 @@
+ * implied. See the License for the specific language governing
+ * rights and limitations under the License.
+ *
+ * Alternatively, the contents of this file may be used under the terms of
+ * either of the GNU General Public License Version 2 or later (the "GPL"),
+ * or the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
+ * in which case the provisions of the GPL or the LGPL are applicable instead
+ * of those above. If you wish to allow use of your version of this file only
+ * under the terms of either the GPL or the LGPL, and not to allow others to
+ * use your version of this file under the terms of the EPL, indicate your
+ * decision by deleting the provisions above and replace them with the notice
+ * and other provisions required by the GPL or the LGPL. If you do not delete
+ * the provisions above, a recipient may use your version of this file under
+ * the terms of any one of the EPL, the GPL or the LGPL.
+ ***** END LICENSE BLOCK *****/
+package org.jruby.specialized;
+
+import me.qmx.jitescript.CodeBlock;
+import me.qmx.jitescript.JDKVersion;
+import me.qmx.jitescript.JiteClass;
+import org.jruby.ReifiedRubyObject;
+import org.jruby.Ruby;
+import org.jruby.RubyClass;
+import org.jruby.RubyInstanceConfig;
+import org.jruby.compiler.impl.SkinnyMethodAdapter;
+import org.jruby.parser.StaticScope;
+import org.jruby.runtime.DynamicScope;
+import org.jruby.runtime.Helpers;
+import org.jruby.runtime.ObjectAllocator;
+import org.jruby.runtime.builtin.IRubyObject;
+import org.jruby.util.ClassDefiningClassLoader;
+import org.jruby.util.CodegenUtils;
+import org.jruby.util.OneShotClassLoader;
+import org.jruby.util.collections.NonBlockingHashMapLong;
+import org.objectweb.asm.ClassWriter;
+import org.objectweb.asm.Label;
+import org.objectweb.asm.tree.LabelNode;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.util.Set;
+
+import static org.jruby.util.CodegenUtils.ci;
+import static org.jruby.util.CodegenUtils.p;
+import static org.jruby.util.CodegenUtils.sig;
+import static org.objectweb.asm.Opcodes.*;
+
+/**
+ * A code generator for Ruby objects, to map known instance variables into fields.
+ */
+public class RubyObjectSpecializer {
+
+    private static ObjectAllocator getClassFromSize(int size) {
+        return specializedFactories.get(size);
+    }
+
+    private static final NonBlockingHashMapLong<ObjectAllocator> specializedFactories = new NonBlockingHashMapLong<>();
+
+    private static ClassDefiningClassLoader CDCL = new OneShotClassLoader(Ruby.getClassLoader());
+
+    public static ObjectAllocator specializeForVariables(RubyClass klass, Set<String> foundVariables) {
+        int size = foundVariables.size();
+        ObjectAllocator allocator = getClassFromSize(size);
+
+        if (allocator != null) return allocator;
+
+        final String clsPath = "org/jruby/specialize/RubyObject" + size;
```

Comment:
I agree this generated code should go into org.jruby.gen. The "org.jruby.specialize" was a typo on my part; it would have been specialized, which I think we will keep to separate the hand-written pieces from the generated stuff.